File: src/main/java/me/kyleseven/consolereader/logreader/LogAppenderManager.java
package me.kyleseven.consolereader.logreader;

import org.apache.logging.log4j.core.Logger;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.UUID;

public class LogAppenderManager {
    private static HashMap<UUID, LogAppender> logAppenders;
    private static Logger logger;

    public static void setup(Logger logger) {
        LogAppenderManager.logAppenders = new HashMap<>();
        LogAppenderManager.logger = logger;
    }

    public static void startReading(Player player) {
        LogAppender appender = new LogAppender(player);
        logAppenders.put(player.getUniqueId(), appender);
        logger.addAppender(appender);
    }

    public static void startReadingTemp(Player player, Long milliseconds) {
        LogAppender appender = new LogAppender(player);
        logger.addAppender(appender);
        new Thread(() -> {
            try {
                Thread.sleep(milliseconds);
                logger.removeAppender(appender);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
    }

    public static void stopReading(Player player) {
        logger.removeAppender(logAppenders.get(player.getUniqueId()));
        logAppenders.remove(player.getUniqueId());
    }

    public static void stopReadingTemp(Player player) {
        logger.removeAppender(logAppenders.get(player.getUniqueId()));
    }

    public static void stopReadingAll() {
        for (LogAppender appender : logAppenders.values()) {
            logger.removeAppender(appender);
        }

        logAppenders.clear();
    }

    public static boolean isReading(Player player) {
        return logAppenders.containsKey(player.getUniqueId());
    }
}

File: src/main/kotlin/me/kyleseven/consolereader/logappender/LogAppenderManager.kt
package me.kyleseven.consolereader.logappender

import org.apache.logging.log4j.core.Logger
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*
import kotlin.collections.HashMap
import kotlin.concurrent.thread

object LogAppenderManager {
    private lateinit var logger: Logger
    private lateinit var logAppenders: HashMap<UUID, LogAppender?>

    fun setup(logger: Logger) {
        this.logger = logger
        logAppenders = hashMapOf()
    }

    fun getReadingPlayerUUIDs(): List<UUID> {
        return logAppenders.keys.toList()
    }

    fun isReading(player: OfflinePlayer): Boolean {
        return logger.appenders.containsKey("ConsoleReader-${player.uniqueId}") || logAppenders.containsKey(player.uniqueId)
    }

    fun startReading(player: OfflinePlayer) {
        if (player.isOnline) {
            val appender = LogAppender(player as Player)
            logger.addAppender(appender)
            logAppenders[player.uniqueId] = appender
        } else {
            logAppenders[player.uniqueId] = null
        }
    }

    // Requires online player
    fun startReadingTemp(player: Player, milliseconds: Long) {
        val appender = LogAppender(player)
        logger.addAppender(appender)
        thread {
            Thread.sleep(milliseconds)
            logger.removeAppender(appender)
        }
    }

    fun stopReading(player: OfflinePlayer) {
        if (logAppenders[player.uniqueId] != null) {
            logger.removeAppender(logAppenders[player.uniqueId])
        }
        logAppenders.remove(player.uniqueId)
    }

    // Requires online player
    fun stopReadingTemp(player: Player) {
        logger.removeAppender(logAppenders[player.uniqueId])
    }

    fun stopReadingAll() {
        for (app in logAppenders.values) {
            if (app != null) {
                logger.removeAppender(app)
            }
        }

        logAppenders.clear()
    }
}